sphinxcontrib/needs/api/configuration.py

```diff
@@ -118,7 +124,13 @@ def my_function(app, need, needs, *args, **kwargs):
     register_func(function, name)
 
 
-def add_warning(app, name, function=None, filter_string=None):
+# 'Need' is untyped, so we temporarily use 'Any' here
+WarningCheck = Callable[[Any, SphinxLoggerAdapter], bool]
+
+
+def add_warning(
+    app: Sphinx, name: str, function: Optional[WarningCheck] = None, filter_string: Optional[str] = None
+) -> None:
```

Comment:
hey @danwos can you check my work here? the function signature (and purpose) of the 'warning check' function don't seem to be documented anywhere, but there are examples, which is where i've tried to infer the signature from.

I've used `Any` for the 'need' since the need is currently untyped anyway.

Reply:
on an unrelated note, this function can accept a `function` or a `filter_string` arg, and has checks to ensure that there is at least one set, and not both. You could remove these checks by changing the signature to

```python
def add_warning(app: Sphinx, name: str, check: Union[WarningCheck, str]) -> None:
```

and then doing an `isinstance` check to see which one was used.

that's a breaking change, so doesn't belong in this PR

Reply:
This looks good.
And oh yes, docs are totally missing for the warning functions.

It's really great to see typing annotation being added. Especially as I'm still learning a lot of tricks from you.
So thanks for all the free lessons and for sure also for the PRs :+1: